plugins/xlsxreader.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Wrapper for xlsxio_read library functions.
+ */
+
+#define BUILD_XLSXIO_STATIC
+#include <xlsxio_read.h>
+
+#include "DataFuncs.h"
+#include "LuaTools.h"
+#include "PluginManager.h"
+#include "PluginStatics.h"
+
+using namespace DFHack;
+
+DFHACK_PLUGIN("xlsxreader");
+
+// returns NULL on error
+xlsx_file_handle* open_xlsx_file(std::string filename) {
+    return new xlsx_file_handle(xlsxioread_open(filename.c_str()));
+}
+
+void close_xlsx_file(xlsx_file_handle *file_handle) {
+    CHECK_NULL_POINTER(file_handle);
+    xlsxioread_close(file_handle->handle);
+    delete(file_handle);
+}
+
+// returns XLSXIOReaderSheet object or NULL on error
+xlsx_sheet_handle* open_sheet(xlsx_file_handle *file_handle,
```

Comment:
I haven't done a lot of debugging yet, but this function segfaults (due to a null `handle` passed to `xlsxioread_sheet_open()`) if the xlsx file failed to open, due to it not existing (which happened to me when testing originally) or the user lacking permission to read it. Quickfort seems to have pretty robust checks for the former, i.e. not opening files that don't exist, but the latter might cause issues. I haven't gotten `quickfort list` to segfault yet (it prints the `no sheet with a #mode marker detected` message for non-readable files), but other user code might.

Is there a good place to handle this? Maybe if `xlsxioread_open` returns null, that should throw a Lua error in `open_xlsx_file`? I think that should prevent Lua code from obtaining a null handle in the first place.

Reply:
I think this can be solved with better use of CHECK_NULL_POINTER. I'm currently using it to verify my *wrapper*, but not the inner handle itself. Let me test that a bit and make sure that protects successfully against segfaults

Reply:
the problem was that I didn't update my NULL detection when I switched to using the _identity-compliant wrappers. I wasn't able to induce any NULL-related segfaults with the updated checks. Thank you for catching this!